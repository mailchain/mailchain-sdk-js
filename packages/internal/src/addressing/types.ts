import { ProtocolType } from '../protocols';

export type MailchainAddress = {
	value: string;
	protocol: ProtocolType;
	domain: string;
};
