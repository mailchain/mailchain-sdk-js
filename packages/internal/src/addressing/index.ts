export * from './encoding';
export * from './test.constants';
export * from './formatMailLike';
