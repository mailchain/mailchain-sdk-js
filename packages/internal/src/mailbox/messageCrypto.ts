import { EncryptedContent, ED25519ExtendedPrivateKey } from '@mailchain/crypto';
import { KeyRing } from '@mailchain/keyring';
import { deserialize, serialize, CHUNK_LENGTH_1MB, decryptPayload, encryptPayload } from '../transport/serialization';
import { Payload } from '../transport';
import { SerializablePayloadHeadersImpl } from '../transport/payload/headersSerialize';

/**
 * Used for encryption and decryption of the full message content.
 */
export type MessageCrypto = {
	encrypt: (payload: Payload) => Promise<EncryptedContent>;
	decrypt: (content: EncryptedContent) => Promise<Payload>;
};

export function createMailchainMessageCrypto(keyRing: KeyRing): MessageCrypto {
	const inboxKey = ED25519ExtendedPrivateKey.fromPrivateKey(keyRing.rootInboxKey());
	const headersSerializer = new SerializablePayloadHeadersImpl();

	const encrypt: MessageCrypto['encrypt'] = async (payload) => {
		const headers = headersSerializer.serialize(payload.Headers);
		const encryptedPayload = await encryptPayload(headers, payload.Content, inboxKey, CHUNK_LENGTH_1MB);
		const serializedContent = serialize(encryptedPayload);
		return new Uint8Array(serializedContent);
	};

	const decrypt: MessageCrypto['decrypt'] = async (serializedContent) => {
		const encryptedPayload = deserialize(Buffer.from(serializedContent));
		const { headers, content } = await decryptPayload(encryptedPayload, inboxKey);

		return {
			Content: content,
			Headers: headersSerializer.deserialize(headers),
		};
	};

	return { encrypt, decrypt };
}
