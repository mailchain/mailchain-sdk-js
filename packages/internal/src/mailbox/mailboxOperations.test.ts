import { formatAddress, parseNameServiceAddress } from '@mailchain/addressing';
import {
	GetMailboxOverviewResponseBody,
	GetMessageResponseBody,
	GetMessagesInViewResponseBody,
	InboxApi,
	PostPayloadResponseBody,
	PutEncryptedMessageRequestBodyFolderEnum,
} from '@mailchain/api';
import { KindNaClSecretKey, publicKeyToBytes } from '@mailchain/crypto';
import { AliceED25519PrivateKey } from '@mailchain/crypto/ed25519/test.const';
import { EncodingTypes, decodeBase64, encodeBase64, encodeHex } from '@mailchain/encoding';
import { KeyRing } from '@mailchain/keyring';
import { sha256 } from '@noble/hashes/sha256';
import { AxiosResponse } from 'axios';
import { mock } from 'jest-mock-extended';
import { createMimeMessage } from '../formatters/generate';
import { MailboxRuleEngine } from '../mailboxRuleEngine';
import * as protoInbox from '../protobuf/inbox/inbox';
import { dummyMailData } from '../test.const';
import { Payload } from '../transport';
import { AliceAccountMailbox, AliceWalletMailbox, BobAccountMailbox } from '../user/test.const';
import { AddressesHasher } from './addressHasher';
import { MailboxOperations, MailchainMailboxOperations } from './mailboxOperations';
import { createMailchainMessageCrypto } from './messageCrypto';
import { createMailchainMessageIdCreator } from './messageId';
import { AddressMatch, MessageMailboxOwnerMatcher } from './messageMailboxOwnerMatcher';
import { MessagePreviewMigrationRule } from './migrations';
import { SystemMessageLabel } from './types';
import { UserMailboxHasher } from './userMailboxHasher';

describe('mailbox', () => {
	const keyRing = KeyRing.fromPrivateKey(AliceED25519PrivateKey);
	const messagePreviewCrypto = keyRing.inboxKey();
	const messageCrypto = createMailchainMessageCrypto(keyRing);
	const mockRuleEngine = mock<MailboxRuleEngine>();

	const mockAddressHasher: AddressesHasher = (addresses) =>
		Promise.resolve(
			addresses.reduce((result, address) => {
				result.set(address, [
					{ type: 'identity-key', hash: Buffer.from(address) },
					{ type: 'username', hash: Buffer.from(address) },
				]);
				return result;
			}, new Map()),
		);
	const mockUserMailboxHasher: UserMailboxHasher = (mailbox) => Promise.resolve(sha256(mailbox.identityKey.bytes));
	const mockInboxApi = mock<InboxApi>();
	const mockOwnerMatcher = mock<MessageMailboxOwnerMatcher>();
	let payload: Payload;

	const msg1Preview = protoInbox.preview.MessagePreview.create({
		from: 'Bob',
		subject: 'Hey Alice!',
		owner: 'ownerAddress',
		mailbox: publicKeyToBytes(AliceAccountMailbox.identityKey),
		snippet: 'Message from Bob to Alice',
		hasAttachment: true,
		timestamp: new Date('2022-6-6').getTime() / 1000,
		to: ['0x4321@ethereum.mailchain.local', '0xabcd@ethereum.mailchain.local'],
	});
	const msg2Preview = { ...msg1Preview, snippet: 'Another message from Bob to Alice' };

	const dummyMigration: MessagePreviewMigrationRule = {
		shouldApply: ({ version }) => Promise.resolve(version === 1),
		apply: (data) =>
			Promise.resolve({
				messagePreview: data.messagePreview,
				version: 2,
			}),
	};

	let mailboxOperations: MailboxOperations;

	const dateOffset = 999;
	beforeEach(() => {
		jest.clearAllMocks();

		mockRuleEngine.apply.mockImplementation((params) => {
			return Promise.resolve({
				...params,
				message: {
					...params.message,
					systemLabels: [...params.message.systemLabels, 'rule-engine' as SystemMessageLabel],
				},
			});
		});

		mockOwnerMatcher.withMessageIdentityKeys.mockReturnValue(mockOwnerMatcher);
		mailboxOperations = new MailchainMailboxOperations(
			mockInboxApi,
			messagePreviewCrypto,
			messageCrypto,
			mockOwnerMatcher,
			mockAddressHasher,
			createMailchainMessageIdCreator(keyRing),
			mockUserMailboxHasher,
			dateOffset,
			dummyMigration,
			mockRuleEngine,
		);
	});

	beforeAll(async () => {
		payload = {
			Headers: {
				Origin: keyRing.accountMessagingKey().publicKey,
				ContentSignature: new Uint8Array([1, 3, 3, 7]),
				Created: new Date('2022-6-14'),
				ContentLength: 1337,
				ContentType: 'message/x.mailchain',
				ContentEncoding: EncodingTypes.Base64,
				ContentEncryption: KindNaClSecretKey,
			},
			Content: Buffer.from((await createMimeMessage(dummyMailData, new Map())).original),
		};
	});

	it('should return message preview for message', async () => {
		const msg1EncryptedPreview = await messagePreviewCrypto.encrypt(
			protoInbox.preview.MessagePreview.encode(msg1Preview).finish(),
		);
		mockInboxApi.getMessage.mockResolvedValue({
			data: {
				message: {
					kind: 'mail',
					messageId: encodeHex(Uint8Array.from([1])),
					encryptedPreview: encodeBase64(msg1EncryptedPreview),
					systemLabels: ['inbox', 'unread'],
				},
			},
		} as AxiosResponse<GetMessageResponseBody>);

		const message = await mailboxOperations.getMessage('messageId');

		expect(mockInboxApi.getMessage.mock.calls[0][0]).toEqual('messageId');
		expect(message).toEqual({
			...msg1Preview,
			kind: 'mail',
			mailbox: AliceAccountMailbox.identityKey,
			messageId: encodeHex(Uint8Array.from([1])),
			isRead: false,
			timestamp: new Date(msg1Preview.timestamp * 1000),
			systemLabels: ['inbox', 'unread'],
		});
	});

	const messageViewsTests = [
		['getInboxMessages', 'getMessagesInInboxView'],
		['getArchivedMessages', 'getMessagesInArchivedView'],
		['getStarredMessages', 'getMessagesInStarredView'],
		['getTrashMessages', 'getMessagesInTrashView'],
		['getUnreadMessages', 'getMessagesInUnreadView'],
		['getSentMessages', 'getMessagesInSentView'],
		['getSpamMessages_unstable', 'getMessagesInSpamView'],
	] as const;

	test.each(messageViewsTests)(
		'should return correct messages for %p using %p api method',
		async (mailboxMethod, apiMethod) => {
			// Given
			const msg1EncryptedPreview = await messagePreviewCrypto.encrypt(
				protoInbox.preview.MessagePreview.encode(msg1Preview).finish(),
			);
			const msg2EncryptedPreview = await messagePreviewCrypto.encrypt(
				protoInbox.preview.MessagePreview.encode(msg2Preview).finish(),
			);
			mockInboxApi[apiMethod].mockResolvedValue({
				data: {
					messages: [
						{
							kind: 'mail',
							messageId: encodeHex(Uint8Array.from([1])),
							encryptedPreview: encodeBase64(msg1EncryptedPreview),
							systemLabels: ['starred'],
						},
						{
							kind: 'mail',
							messageId: encodeHex(Uint8Array.from([2])),
							encryptedPreview: encodeBase64(msg2EncryptedPreview),
							systemLabels: ['starred'],
						},
					],
				},
			} as AxiosResponse<GetMessagesInViewResponseBody>);

			// When
			const messages = await mailboxOperations[mailboxMethod]({
				offset: 10,
				limit: 20,
				userMailboxes: [AliceAccountMailbox],
			});

			// Then
			expect(mockInboxApi[apiMethod]).toHaveBeenCalledTimes(1);
			expect(mockInboxApi[apiMethod]).toHaveBeenCalledWith(undefined, undefined, undefined, 10, 20, [
				'e80f39fd4a3d65d4a6494125ee0ecf279024c8fdd4e670e225485c777349d5a9',
			]);
			expect(messages).toEqual(
				[msg1Preview, msg2Preview].map((msg, index) => ({
					...msg,
					kind: 'mail',
					mailbox: AliceAccountMailbox.identityKey,
					messageId: encodeHex(Uint8Array.from([index + 1])),
					isRead: true,
					timestamp: new Date(msg.timestamp * 1000),
					systemLabels: ['starred'],
					to: ['0x4321@ethereum.mailchain.local', '0xabcd@ethereum.mailchain.local'],
				})),
			);
		},
	);

	it('should decrypt full message body', async () => {
		const encryptedPayload = await messageCrypto.encrypt(payload);
		mockInboxApi.getEncryptedMessageBody.mockResolvedValue({ data: encryptedPayload } as AxiosResponse<object>);

		const message = await mailboxOperations.getFullMessage('messageId');

		expect(message.replyTo).toEqual(dummyMailData.replyTo?.address);
		expect(message.body).toMatchSnapshot('fullMessageBody');
	});

	it('should encrypt and post sent message', async () => {
		const uri = 'messageBodyUri';
		const resourceId = 'resourceId';
		mockInboxApi.postEncryptedMessageBody.mockResolvedValue({
			data: { uri, resourceId },
		} as AxiosResponse<PostPayloadResponseBody>);
		mockInboxApi.putEncryptedMessage.mockResolvedValue({ data: undefined } as AxiosResponse<void>);

		const message = await mailboxOperations.saveSentMessage({
			userMailbox: AliceAccountMailbox,
			payload,
			content: dummyMailData,
		});

		expect(message.messageId).toMatchSnapshot('sent message id');
		expect(mockInboxApi.putEncryptedMessage.mock.calls[0][0]).toEqual(message.messageId);
		const requestBody = mockInboxApi.putEncryptedMessage.mock.calls[0][1];
		const decryptedPreview = protoInbox.preview.MessagePreview.decode(
			await messagePreviewCrypto.decrypt(decodeBase64(requestBody.encryptedPreview)),
		);

		expect(decryptedPreview).toEqual({
			owner: formatAddress(AliceAccountMailbox.aliases[0].address, 'mail'),
			mailbox: publicKeyToBytes(AliceAccountMailbox.identityKey),
			from: dummyMailData.from.address,
			hasAttachment: false,
			snippet:
				'- Lorem ipsum dolor sit amet, consectetuer adipiscing elit. - Aliquam tincidunt mauris eu risus. -',
			subject: dummyMailData.subject,
			timestamp: payload.Headers.Created.getTime() / 1000,
			to: dummyMailData.recipients.map((r) => r.address),
			cc: dummyMailData.carbonCopyRecipients.map((r) => r.address),
			bcc: dummyMailData.blindCarbonCopyRecipients.map((r) => r.address),
		});
		expect(requestBody.kind).toEqual('mail');
		expect(requestBody.version).toEqual(3);
		expect(requestBody.date + dateOffset).toEqual(payload.Headers.Created.getTime() / 1000);
		expect(requestBody.folder).toEqual(PutEncryptedMessageRequestBodyFolderEnum.Outbox);
		expect(requestBody.messageBodyResourceId).toEqual(resourceId);
		expect(requestBody.hashedOwner).toMatchSnapshot('hashedOwner');
		expect(requestBody.hashedFrom).toMatchSnapshot('hashedFrom');
		expect(requestBody.hashedTo).toMatchSnapshot('hashedTo');
		expect(requestBody.mailbox).toEqual(Array.from(sha256(AliceAccountMailbox.identityKey.bytes)));
		expect(
			await messageCrypto.decrypt(mockInboxApi.postEncryptedMessageBody.mock.calls[0][0] as Uint8Array),
		).toEqual(payload);
	});

	it('should encrypt and post received message', async () => {
		const matchedOwners: AddressMatch[] = [
			{ matchBy: 'message-header', address: parseNameServiceAddress(dummyMailData.recipients[0].address) },
			{
				matchBy: 'mailchain-api',
				address: parseNameServiceAddress(dummyMailData.carbonCopyRecipients[0].address),
			},
		];
		mockOwnerMatcher.findMatches.mockResolvedValue(matchedOwners);
		const uri = 'messageBodyUri';
		const resourceId = 'resourceId';
		mockInboxApi.postEncryptedMessageBody.mockResolvedValue({
			data: { uri, resourceId },
		} as AxiosResponse<PostPayloadResponseBody>);
		mockInboxApi.putEncryptedMessage.mockResolvedValue({ data: undefined } as AxiosResponse<void>);

		const messages = await mailboxOperations.saveReceivedMessage({
			receivedTransportPayload: payload,
			userMailbox: AliceWalletMailbox,
		});

		expect(messages).toHaveLength(matchedOwners.length);
		expect(mockInboxApi.putEncryptedMessage).toHaveBeenCalledTimes(matchedOwners.length);
		expect(mockRuleEngine.apply).toHaveBeenCalledTimes(matchedOwners.length);

		for (let index = 0; index < matchedOwners.length; index++) {
			const { address: matchedOwner } = matchedOwners[index];
			expect(messages[index]).toMatchSnapshot(`message id ${formatAddress(matchedOwner, 'mail')}`);
			expect(mockInboxApi.putEncryptedMessage.mock.calls[index][0]).toEqual(messages[index].messageId);
			const requestBody = mockInboxApi.putEncryptedMessage.mock.calls[index][1];
			const decryptedPreview = protoInbox.preview.MessagePreview.decode(
				await messagePreviewCrypto.decrypt(decodeBase64(requestBody.encryptedPreview)),
			);

			expect(decryptedPreview).toEqual({
				from: dummyMailData.from.address,
				owner: formatAddress(matchedOwner, 'mail'),
				mailbox: publicKeyToBytes(AliceWalletMailbox.identityKey),
				hasAttachment: false,
				snippet:
					'- Lorem ipsum dolor sit amet, consectetuer adipiscing elit. - Aliquam tincidunt mauris eu risus. -',
				subject: dummyMailData.subject,
				timestamp: payload.Headers.Created.getTime() / 1000,
				to: dummyMailData.recipients.map((r) => r.address),
				cc: dummyMailData.carbonCopyRecipients.map((r) => r.address),
				bcc: dummyMailData.blindCarbonCopyRecipients.map((r) => r.address),
			});
			expect(requestBody.kind).toEqual('mail');
			expect(requestBody.version).toEqual(3);
			expect(requestBody.mailbox).toEqual(Array.from(sha256(AliceWalletMailbox.identityKey.bytes)));
			expect(requestBody.date + dateOffset).toEqual(payload.Headers.Created.getTime() / 1000);
			expect(requestBody.folder).toEqual(PutEncryptedMessageRequestBodyFolderEnum.Inbox);
			expect(requestBody.messageBodyResourceId).toEqual(resourceId);
			expect(requestBody.hashedOwner).toMatchSnapshot('hashedOwner');
			expect(requestBody.hashedFrom).toMatchSnapshot('hashedFrom');
			expect(requestBody.hashedTo).toMatchSnapshot('hashedTo');
			expect(
				await messageCrypto.decrypt(mockInboxApi.postEncryptedMessageBody.mock.calls[0][0] as Uint8Array),
			).toEqual(payload);
		}
	});

	it('should remove html tags from message snippet', async () => {
		mockOwnerMatcher.findMatches.mockResolvedValue([
			{ matchBy: 'message-header', address: parseNameServiceAddress(dummyMailData.recipients[0].address) },
		]);
		const uri = 'messageBodyUri';
		const resourceId = 'resourceId';
		mockInboxApi.postEncryptedMessageBody.mockResolvedValue({
			data: { uri, resourceId },
		} as AxiosResponse<PostPayloadResponseBody>);
		mockInboxApi.putEncryptedMessage.mockResolvedValue({ data: undefined } as AxiosResponse<void>);

		const payloadWithHtmlPlainText: Payload = {
			...payload,
			Content: Buffer.from(
				(await createMimeMessage({ ...dummyMailData, plainTextMessage: dummyMailData.message }, new Map()))
					.original,
			),
		};
		const messages = await mailboxOperations.saveReceivedMessage({
			receivedTransportPayload: payloadWithHtmlPlainText,
			userMailbox: AliceWalletMailbox,
		});

		expect(messages).toHaveLength(1);
		const message = messages[0];
		expect(message.snippet).toMatchInlineSnapshot(
			`"Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aliquam tincidunt mauris eu risus. Vesti"`,
		);
	});

	const labelsTests = [
		['modifyArchiveMessage', 'archive', true],
		['modifyIsReadMessage', 'unread', false],
		['modifyTrashMessage', 'trash', true],
		['modifyStarredMessage', 'starred', true],
		['modifyUserLabel', 'user-label', true],
	] as const;

	describe.each(labelsTests)('%s', (method, label, shouldPut) => {
		it(`should put ${label} when ${method} is invoked with ${shouldPut}`, async () => {
			if (method === 'modifyUserLabel') {
				await mailboxOperations[method]('messageId', label, shouldPut);
			} else {
				await mailboxOperations[method]('messageId', shouldPut);
			}

			const apiCall = mockInboxApi.putMessageLabel.mock.calls[0];
			expect(apiCall[0]).toEqual('messageId');
			expect(apiCall[1]).toEqual(label);
		});

		it(`should delete ${label} when ${method} is invoked with ${!shouldPut}`, async () => {
			if (method === 'modifyUserLabel') {
				await mailboxOperations[method]('messageId', label, !shouldPut);
			} else {
				await mailboxOperations[method]('messageId', !shouldPut);
			}

			const apiCall = mockInboxApi.deleteMessageLabel.mock.calls[0];
			expect(apiCall[0]).toEqual('messageId');
			expect(apiCall[1]).toEqual(label);
		});
	});

	it('should compute messages overview', async () => {
		mockInboxApi.getMailboxOverview.mockResolvedValue({
			data: {
				mailboxes: [
					{
						labels: [
							{ label: 'inbox', total: 2, unread: 1 },
							{ label: 'starred', total: 3, unread: 2 },
							{ label: 'archived', total: 4, unread: 3 },
						],
						mailbox: encodeHex(await mockUserMailboxHasher(AliceAccountMailbox)),
					},
					{
						labels: [
							{ label: 'inbox', total: 3, unread: 0 },
							{ label: 'starred', total: 4, unread: 1 },
							{ label: 'sent', total: 5, unread: 0 },
						],
						mailbox: encodeHex(await mockUserMailboxHasher(AliceWalletMailbox)),
					},
				],
			},
		} as AxiosResponse<GetMailboxOverviewResponseBody>);

		const overview = await mailboxOperations.getMessagesOverview([
			AliceAccountMailbox,
			AliceWalletMailbox,
			BobAccountMailbox,
		]);

		expect(overview).toMatchSnapshot();
		expect(mockInboxApi.getMailboxOverview).toHaveBeenCalledWith([
			'e80f39fd4a3d65d4a6494125ee0ecf279024c8fdd4e670e225485c777349d5a9',
			'c35dd5d86b8f0170db2b082cef3bca93db9ed95c6c785d4a67cced08b773bcea',
			'd2c8b50395acebb32709406a1102c4aa25f2fefc57d0ed64f58239d1be1da701',
		]);
	});
});
