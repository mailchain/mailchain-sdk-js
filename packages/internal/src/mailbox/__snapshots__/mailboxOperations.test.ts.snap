// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mailbox should decrypt full message body: fullMessageBody 1`] = `
"<ul>
	<li>Lorem ipsum dolor sit amet, consectetuer adipiscing elit.</li>
	<li>Aliquam tincidunt mauris eu risus.</li>
	<li>Vestibulum auctor dapibus neque.</li>
</ul>"
`;

exports[`mailbox should encrypt and post received message: hashedFrom 1`] = `
[
  97,
  108,
  105,
  99,
  101,
  64,
  109,
  97,
  105,
  108,
  99,
  104,
  97,
  105,
  110,
  46,
  116,
  101,
  115,
  116,
]
`;

exports[`mailbox should encrypt and post received message: hashedFrom 2`] = `
[
  97,
  108,
  105,
  99,
  101,
  64,
  109,
  97,
  105,
  108,
  99,
  104,
  97,
  105,
  110,
  46,
  116,
  101,
  115,
  116,
]
`;

exports[`mailbox should encrypt and post received message: hashedOwner 1`] = `
[
  98,
  111,
  98,
  64,
  109,
  97,
  105,
  108,
  99,
  104,
  97,
  105,
  110,
  46,
  116,
  101,
  115,
  116,
]
`;

exports[`mailbox should encrypt and post received message: hashedOwner 2`] = `
[
  48,
  120,
  57,
  50,
  100,
  56,
  102,
  49,
  48,
  50,
  52,
  56,
  99,
  54,
  97,
  51,
  57,
  53,
  51,
  99,
  99,
  51,
  54,
  57,
  50,
  97,
  56,
  57,
  52,
  54,
  53,
  53,
  97,
  100,
  48,
  53,
  100,
  54,
  49,
  101,
  102,
  98,
  64,
  101,
  116,
  104,
  101,
  114,
  101,
  117,
  109,
  46,
  109,
  97,
  105,
  108,
  99,
  104,
  97,
  105,
  110,
  46,
  116,
  101,
  115,
  116,
]
`;

exports[`mailbox should encrypt and post received message: hashedTo 1`] = `
[
  [
    98,
    111,
    98,
    64,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    98,
    111,
    98,
    64,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    48,
    120,
    100,
    53,
    97,
    98,
    52,
    99,
    101,
    51,
    54,
    48,
    53,
    99,
    100,
    53,
    57,
    48,
    100,
    98,
    54,
    48,
    57,
    98,
    54,
    98,
    53,
    99,
    56,
    57,
    48,
    49,
    102,
    100,
    98,
    50,
    101,
    102,
    55,
    102,
    101,
    54,
    64,
    101,
    116,
    104,
    101,
    114,
    101,
    117,
    109,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    48,
    120,
    100,
    53,
    97,
    98,
    52,
    99,
    101,
    51,
    54,
    48,
    53,
    99,
    100,
    53,
    57,
    48,
    100,
    98,
    54,
    48,
    57,
    98,
    54,
    98,
    53,
    99,
    56,
    57,
    48,
    49,
    102,
    100,
    98,
    50,
    101,
    102,
    55,
    102,
    101,
    54,
    64,
    101,
    116,
    104,
    101,
    114,
    101,
    117,
    109,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    116,
    105,
    109,
    64,
    101,
    116,
    104,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    116,
    105,
    109,
    64,
    101,
    116,
    104,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
]
`;

exports[`mailbox should encrypt and post received message: hashedTo 2`] = `
[
  [
    98,
    111,
    98,
    64,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    98,
    111,
    98,
    64,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    48,
    120,
    100,
    53,
    97,
    98,
    52,
    99,
    101,
    51,
    54,
    48,
    53,
    99,
    100,
    53,
    57,
    48,
    100,
    98,
    54,
    48,
    57,
    98,
    54,
    98,
    53,
    99,
    56,
    57,
    48,
    49,
    102,
    100,
    98,
    50,
    101,
    102,
    55,
    102,
    101,
    54,
    64,
    101,
    116,
    104,
    101,
    114,
    101,
    117,
    109,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    48,
    120,
    100,
    53,
    97,
    98,
    52,
    99,
    101,
    51,
    54,
    48,
    53,
    99,
    100,
    53,
    57,
    48,
    100,
    98,
    54,
    48,
    57,
    98,
    54,
    98,
    53,
    99,
    56,
    57,
    48,
    49,
    102,
    100,
    98,
    50,
    101,
    102,
    55,
    102,
    101,
    54,
    64,
    101,
    116,
    104,
    101,
    114,
    101,
    117,
    109,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    116,
    105,
    109,
    64,
    101,
    116,
    104,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    116,
    105,
    109,
    64,
    101,
    116,
    104,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
]
`;

exports[`mailbox should encrypt and post received message: message id 0x92d8f10248c6a3953cc3692a894655ad05d61efb@ethereum.mailchain.test 1`] = `
{
  "bcc": [
    "jane@mailchain.test",
    "maria@mailchain.test",
  ],
  "cc": [
    "0x92d8f10248c6a3953cc3692a894655ad05d61efb@ethereum.mailchain.test",
    "john@mailchain.test",
  ],
  "from": "alice@mailchain.test",
  "hasAttachment": false,
  "isRead": false,
  "mailbox": SECP256K1PublicKey {
    "bytes": Uint8Array [
      2,
      105,
      217,
      8,
      81,
      14,
      53,
      91,
      235,
      29,
      91,
      242,
      223,
      129,
      41,
      229,
      182,
      64,
      30,
      25,
      105,
      137,
      30,
      128,
      22,
      160,
      178,
      48,
      7,
      57,
      187,
      176,
      6,
    ],
    "curve": "secp256k1",
  },
  "messageId": "6ddf55461c21f3fa1a0b33ee2debfd406311ba4777528194e830f4fd63ef52f7",
  "owner": "0x92d8f10248c6a3953cc3692a894655ad05d61efb@ethereum.mailchain.test",
  "snippet": "- Lorem ipsum dolor sit amet, consectetuer adipiscing elit. - Aliquam tincidunt mauris eu risus. -",
  "subject": "ðŸ’Œ Dummy MailData subject ðŸ˜‰",
  "systemLabels": [
    "unread",
    "inbox",
    "rule-engine",
  ],
  "timestamp": 2022-06-14T00:00:00.000Z,
  "to": [
    "bob@mailchain.test",
    "0xd5ab4ce3605cd590db609b6b5c8901fdb2ef7fe6@ethereum.mailchain.test",
    "tim@eth.mailchain.test",
  ],
}
`;

exports[`mailbox should encrypt and post received message: message id bob@mailchain.test 1`] = `
{
  "bcc": [
    "jane@mailchain.test",
    "maria@mailchain.test",
  ],
  "cc": [
    "0x92d8f10248c6a3953cc3692a894655ad05d61efb@ethereum.mailchain.test",
    "john@mailchain.test",
  ],
  "from": "alice@mailchain.test",
  "hasAttachment": false,
  "isRead": false,
  "mailbox": SECP256K1PublicKey {
    "bytes": Uint8Array [
      2,
      105,
      217,
      8,
      81,
      14,
      53,
      91,
      235,
      29,
      91,
      242,
      223,
      129,
      41,
      229,
      182,
      64,
      30,
      25,
      105,
      137,
      30,
      128,
      22,
      160,
      178,
      48,
      7,
      57,
      187,
      176,
      6,
    ],
    "curve": "secp256k1",
  },
  "messageId": "793dd5d44e166bc50051c959b7668beabfd374fc1ab15ba5601a8fdc566806ac",
  "owner": "bob@mailchain.test",
  "snippet": "- Lorem ipsum dolor sit amet, consectetuer adipiscing elit. - Aliquam tincidunt mauris eu risus. -",
  "subject": "ðŸ’Œ Dummy MailData subject ðŸ˜‰",
  "systemLabels": [
    "unread",
    "inbox",
    "rule-engine",
  ],
  "timestamp": 2022-06-14T00:00:00.000Z,
  "to": [
    "bob@mailchain.test",
    "0xd5ab4ce3605cd590db609b6b5c8901fdb2ef7fe6@ethereum.mailchain.test",
    "tim@eth.mailchain.test",
  ],
}
`;

exports[`mailbox should encrypt and post sent message: hashedFrom 1`] = `
[
  97,
  108,
  105,
  99,
  101,
  64,
  109,
  97,
  105,
  108,
  99,
  104,
  97,
  105,
  110,
  46,
  116,
  101,
  115,
  116,
]
`;

exports[`mailbox should encrypt and post sent message: hashedOwner 1`] = `
[
  97,
  108,
  105,
  99,
  101,
  64,
  109,
  97,
  105,
  108,
  99,
  104,
  97,
  105,
  110,
  46,
  116,
  101,
  115,
  116,
]
`;

exports[`mailbox should encrypt and post sent message: hashedTo 1`] = `
[
  [
    98,
    111,
    98,
    64,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    98,
    111,
    98,
    64,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    48,
    120,
    100,
    53,
    97,
    98,
    52,
    99,
    101,
    51,
    54,
    48,
    53,
    99,
    100,
    53,
    57,
    48,
    100,
    98,
    54,
    48,
    57,
    98,
    54,
    98,
    53,
    99,
    56,
    57,
    48,
    49,
    102,
    100,
    98,
    50,
    101,
    102,
    55,
    102,
    101,
    54,
    64,
    101,
    116,
    104,
    101,
    114,
    101,
    117,
    109,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    48,
    120,
    100,
    53,
    97,
    98,
    52,
    99,
    101,
    51,
    54,
    48,
    53,
    99,
    100,
    53,
    57,
    48,
    100,
    98,
    54,
    48,
    57,
    98,
    54,
    98,
    53,
    99,
    56,
    57,
    48,
    49,
    102,
    100,
    98,
    50,
    101,
    102,
    55,
    102,
    101,
    54,
    64,
    101,
    116,
    104,
    101,
    114,
    101,
    117,
    109,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    116,
    105,
    109,
    64,
    101,
    116,
    104,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
  [
    116,
    105,
    109,
    64,
    101,
    116,
    104,
    46,
    109,
    97,
    105,
    108,
    99,
    104,
    97,
    105,
    110,
    46,
    116,
    101,
    115,
    116,
  ],
]
`;

exports[`mailbox should encrypt and post sent message: sent message id 1`] = `"8abcfc68fff062b5776787aee9a56f0d6ba84c5c8e7a977309825b76fbe86116"`;
